In [2]:
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from numpy.random import normal

## Maximum Likelihood Estimation

### Likelihood Function

To understand how to fit a distribution from data points observed, we need to have some assumption about the underlying distribution. For example, given the height (or weight) of 100 individuals, one might want to fit a multivariate (perhaps correlated) gaussian to the data set.

How do we accomplish this? Let's cover some basic definitions

> **Definition**: Likelihood
Likelihood refers to the **value** of the probability density for a certain value of the observed random variable.

> **Definition**: Point Estimation is the process of finding the best fitting parameter of a distribution, given the data observed.

Let $f(\mu, \sigma; x_i)$ be the value of the probability density at point $x$, given the mean $\mu$, and standard deviation $\sigma$. Assume $X \sim N(\mu, \sigma)$, therefore, we know that the likelihood is:
$$f(\mu, \sigma; x_i) = \frac{1}{\sqrt{2 \pi \sigma^2}} \exp{\left( -\frac{(x_i - \mu)^2}{2 \sigma^2}\right)}$$
for datapoint $x_i$.

If we have $n$ datapoints: $x_1, x_2, \dots x_n$, then the joint probability of these independent observations is:
$$L(\mu, \sigma; \vec{x}) \equiv f(\mu, \sigma; x_1, x_2, \dots , x_n) = \prod_{i=1}^n  \frac{1}{\sqrt{2 \pi \sigma^2}} \exp{\left( -\frac{(x_i - \mu)^2}{2 \sigma^2}\right)}$$

We call this the likelihood function $L$

How do we maximize this $L$ with respect to parameters $\mu, \sigma$? 

We can take a derivative and find the maximum. In reality, taking a derivative of the **log likelihood** is easier since a derivative of a sum is easier than a derivative of a product!

### Log Likelihood Function


We can solve the following system for $\mu, \sigma$: 

$$\ln{L(\mu, \sigma; \vec{x})} = \sum_{i=1}^n -\ln{\sigma} -\frac{(x_i - \mu)^2}{2 \sigma^2}$$
$$\frac{\partial \ln L}{\partial \mu} = 0$$
$$\frac{\partial \ln L}{\partial \sigma} = 0$$

Or equivalently,
> $$\arg \max_{\mu, \sigma} \ln L$$

The analytical solution as one might expect is:
> $$\hat{\mu} = \frac{1}{n} \sum_{i=1}^n x_i$$
$$\hat{\sigma} = \frac{1}{n} \sum_{i=1}^n (x-\hat{\mu})^2$$


Let's try to do this numerically to learn something new:

- we first generate a sample dataset $x$ sampled from a normal distribution $N(0, 20)$
- then we maximuize the log likelihood by minimizing -1 * LL

In [9]:
x = normal(0, 20, 1000)

In [10]:
def log_likelihood(p):
    mu, sigma = p[0], p[1]
    ll = (-np.log(sigma) - (x - mu)**2 / (2 * sigma**2)).sum()
    return ll * -1

In [11]:
from scipy.optimize import minimize

In [13]:
minimize(log_likelihood, x0 = [0, 1], method='L-BFGS-B')

      fun: 3487.0916733110157
 hess_inv: <2x2 LbfgsInvHessProduct with dtype=float64>
      jac: array([5.91171556e-04, 9.09494627e-05])
  message: 'CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH'
     nfev: 60
      nit: 19
     njev: 20
   status: 0
  success: True
        x: array([-0.26202289, 19.8279443 ])

The result indicate a $\mu \approx -0.26, \sigma \approx 20$, which is quite close to the answer.